### PR TITLE
Expand `@static` docstring

### PR DESCRIPTION
**Pull request by Sbozzolo**
When I first learned about `@static`, I found its docstring not very clear or helpful. For example, I didn't understand that I had to use a conditional (yes, once you understand what `@static` does, it is obvious, but the reason I was reading the docstring is precisely because I didn't know how to use it). I also found the meaning of `@static foo <&&,||> bar` rather unclear. 

I rephrased the docstring to something that I think is clearer and more helpful.

### CODE REVIEW

File: base/osutils.jl [hunk 1]
```diff
@@ -5,11 +5,18 @@
...
-`@static if Sys.isapple() foo end` and `@static foo <&&,||> bar` are also valid syntax.
+This is useful in cases where a construct would be invalid in some cases, such as a `ccall`
+to a os-dependent function, or macros defined in packages that are not imported.
+
+`@static` requires a conditional. The conditional can be in an `if` statement, a ternary
+operator, or `&&`\`||`.
+
+# Example
+
+Suppose we want to parse an expression `expr` that is only on MacOS. We could solve this
```
Comment (KristofferC) [inline base/osutils.jl:16]: > Suppose we want to parse an expression `expr` that is only on MacOS.

I don't really understand this.

**Comment by LilithHafner** [comment]
Thank you for catching starting a fix for this!

The docstring, as it was originally written, is not correct. The Julia execution order is

1. Parsing
2. Macro expansion
3. Lowering
4. Execution (an umbrella term for everything that follows)

The `@static` macro, being a macro, runs at the second phase. Consequently, it does not "Partially evaluate an expression at parse time."

What it actually does is evaluate the conditional (recursively macro expand, lower, and execute) and then, depending on the resulting value, returns only one (or zero) branch of the sub expression, deleting the rest of the branches before they are macro-expanded, lowered, or executed.

The key difference between that and normal conditionals is in the macro expansion and lowering phases.

Another quirk you should know about is that `ccall(blah, blah, blah)` is not a function call. This is strange and unexpected, and appears nowhere in the `ccall` docstring. But notice:

```julia
julia> ccall
ERROR: UndefVarError: `ccall` not defined in `Main`
Suggestion: check for spelling errors or missing imports.

julia> ccall(:jl_get_current_task, Ref{Task}, ()) # ccall is not defined, so this shouldn't work, right.....????
Task (runnable, started) @0x0000ffff19b20200 # Ha! ccall is special and confusing.
```

You can also see that it lowers differently than real functions

```julia
julia> Meta.lower(Main, :(foo(blah, blah, blah)))
:($(Expr(:thunk, CodeInfo(
    @ none within `top-level scope`
1 ─ %1 = foo
│   %2 = blah
│   %3 = blah
│   %4 = blah
│   %5 = (%1)(%2, %3, %4)
└──      return %5
))))

julia> Meta.lower(Main, :(ccall(blah, blah, blah)))
:($(Expr(:error, "ccall argument types must be a tuple; try \"(T,)\"")))

```

So, ccall magic happens, at least in part, during lowering which means it can be skipped by `@static` while it cannot be skipped by ordinary conditionals.

Aside, not relevant to this PR: Hopefully someday we can fix this quirk. In the meantime, a PR documenting the strange behavior of ccall would also be welcome, though ideally phrased in such a way as to allow ccall to become more like an ordinary function in the future

---

Does that help clarify what this macro does?

File: base/osutils.jl (hunk 1 above)
Comment (Sbozzolo) [reply base/osutils.jl]: Oops, I lost a `valid`

**Comment by Sbozzolo** [comment]
> Does that help clarify what this macro does?

Thank you very much for providing this comprehensive explanation! It definetely helps :)

I updated the docstring adding more information from @LilithHafner's message.

**Review by LilithHafner — APPROVED** [review]
Thanks! This is a significant improvement over the existing docstring.

File: base/osutils.jl [hunk 2]
```diff
@@ -3,13 +3,23 @@
...
-Partially evaluate an expression at parse time.
+Partially evaluate an expression at macro expansion time.
 
-For example, `@static Sys.iswindows() ? foo : bar` will evaluate `Sys.iswindows()` and insert
-either `foo` or `bar` into the expression.
-This is useful in cases where a construct would be invalid on other platforms,
-such as a `ccall` to a non-existent function.
-`@static if Sys.isapple() foo end` and `@static foo <&&,||> bar` are also valid syntax.
+This is useful in cases where a construct would be invalid in some cases, such as a `ccall`
+to a os-dependent function, or macros defined in packages that are not imported.
```
Comment (giordano) [inline base/osutils.jl:9]:
```suggestion
to an OS-dependent function, or macros defined in packages that are not imported.
```

File: base/osutils.jl [hunk 3]
```diff
@@ -3,13 +3,23 @@
...
+
+`@static` requires a conditional. The conditional can be in an `if` statement, a ternary
+operator, or `&&`\`||`. The conditional is evaluated by recursively expanding macros,
+lowering and executing the resulting expressions. Then, the matching branch (if any) is
+returned. All the other branches of the conditional are deleted before they are
+macro-expanded (and lowered or executed).
+
+# Example
+
+Suppose we want to parse an expression `expr` that is valid only on MacOS. We could solve
```
Comment (giordano) [inline base/osutils.jl:19]:
```suggestion
Suppose we want to parse an expression `expr` that is valid only on macOS. We could solve
```

File: base/osutils.jl [hunk 4]
```diff
@@ -3,13 +3,23 @@
...
+this problem using `@static` with `@static if Sys.isapple() expr end`. In case we had
+`expr_apple` for MacOS and `expr_others` for the other operating systems, the solution with
```
Comment (giordano) [inline base/osutils.jl:21]:
```suggestion
`expr_apple` for macOS and `expr_others` for the other operating systems, the solution with
```

**Comment by LilithHafner** [comment]
Oops! I only saw your comments when the page refreshed after I clicked merge. Sorry!

**Pull request merged**